drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,1472 @@
+{
+	int ret = 0;
+	uint32_t reg;
+	const struct mmc_rcar_cfg *cfg;
+	struct mmc_rcar_data *data;
+
+	if (!dev || !dev->config) {
+		return -EINVAL;
+	}
+
+	cfg = dev->config;
+	data = dev->data;
+
+	ret = rcar_mmc_poll_reg_flags_check_err(cfg, RCAR_MMC_INFO2,
+				RCAR_MMC_INFO2_CBSY, 0, false);
+	if (ret) {
+		return -ETIMEDOUT;
+	}
+
+	/* soft reset of the host */
+	reg = rcar_mmc_read_reg32(cfg, RCAR_MMC_SOFT_RST);
+	reg &= ~RCAR_MMC_SOFT_RST_RSTX;
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_SOFT_RST, reg);
+	reg |= RCAR_MMC_SOFT_RST_RSTX;
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_SOFT_RST, reg);
+
+	rcar_mmc_reset_and_mask_irqs(cfg);
+
+#ifdef CONFIG_RCAR_MMC_DMA_SUPPORT
+	BUILD_ASSERT(0, "Add reset of DMAC");
+#endif
+
+	data->ddr_mode = 0;
+
+	data->host_io.bus_width = SDHC_BUS_WIDTH4BIT;
+	data->host_io.clock = data->props.f_max >> 7;
+	data->host_io.timing = SDHC_TIMING_LEGACY;
+
+	return 0;
+}
+
+/**
+ * @brief SD Clock (SD_CLK) Output Control Enable
+ *
+ * @note in/out parameters should be checked by a caller function.
+ *
+ * @param cfg MMC driver configuration
+ * @param enable
+ *          false: SD_CLK output is disabled. The SD_CLK signal is fixed 0.
+ *          true:  SD_CLK output is enabled.
+ *
+ * @retval none
+ */
+static void rcar_mmc_enable_clock(const struct mmc_rcar_cfg *cfg, bool enable)
+{
+	uint32_t mmc_clk_ctl = rcar_mmc_read_reg32(cfg, RCAR_MMC_CLKCTL);
+
+	if (enable == true) {
+		mmc_clk_ctl &= ~RCAR_MMC_CLKCTL_OFFEN;
+		mmc_clk_ctl |= RCAR_MMC_CLKCTL_SCLKEN;
+	} else {
+		mmc_clk_ctl |= RCAR_MMC_CLKCTL_OFFEN;
+		mmc_clk_ctl &= ~RCAR_MMC_CLKCTL_SCLKEN;
+	}
+
+	/* Do not change the values of these bits
+	 * when the CBSY bit in SD_INFO2 is 1
+	 */
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_CLKCTL, mmc_clk_ctl);
+
+	/* SD spec recommends at least 1 ms of delay */
+	k_msleep(1);
+}
+
+/**
+ * @brief Convert SDHC response to Renesas MMC response
+ *
+ * Function performs a conversion from SDHC response to Renesas MMC
+ * CMD register response.
+ *
+ * @note in/out parameters should be checked by a caller function.
+ *
+ * @param response_type SDHC response type without SPI flags
+ *
+ * @retval positiv number (partial configuration of CMD register) on
+ *         success, negative errno code otherwise
+ */
+static int32_t rcar_mmc_convert_sd_to_mmc_resp(uint32_t response_type)
+{
+	uint32_t mmc_resp = 0U;
+
+	switch (response_type) {
+	case SD_RSP_TYPE_NONE:
+		mmc_resp = RCAR_MMC_CMD_RSP_NONE;
+		break;
+	case SD_RSP_TYPE_R1:
+	case SD_RSP_TYPE_R5:
+	case SD_RSP_TYPE_R6:
+	case SD_RSP_TYPE_R7:
+		mmc_resp = RCAR_MMC_CMD_RSP_R1;
+		break;
+	case SD_RSP_TYPE_R1b:
+	case SD_RSP_TYPE_R5b:
+		mmc_resp = RCAR_MMC_CMD_RSP_R1B;
+		break;
+	case SD_RSP_TYPE_R2:
+		mmc_resp = RCAR_MMC_CMD_RSP_R2;
+		break;
+	case SD_RSP_TYPE_R3:
+	case SD_RSP_TYPE_R4:
+		mmc_resp = RCAR_MMC_CMD_RSP_R3;
+		break;
+	default:
+		LOG_ERR("unknown response type 0x%08x",
+			response_type);
+		return -EINVAL;
+	}
+
+	if ((int32_t)mmc_resp < 0) {
```

Comment:
really simple, for some special conditions it is happened for some of them not, anyone can extend anything -- not only switch
and yes, I can use assert instead of warning, I'll add it

thanks for comment!